README
Python Client for UIDAI authentication service
=============================================

Status
------

1. Skeleton code (not production-ready) 
2. Simplest possible XSD-compliant XML is generated for request and response 
3. Incoming XML can be validated and 'objectified for both 
4. Outgoing XML can be signed (and verified using p12 file)
5. Has support for generation of Pid XML (biometric and demographic)
6. Session key automatically generated and 
7. Encryption of session key using UID certificate
(and encryption/decryption testing using public.p12) 
8. Support for license key in the config file

To do
-----

1. Look through the GeoAmida implementation to see the differences
in implementation - mostly done 
2. Figure out how to invoke xmlDocDumpFormatMemory instead of 
lxml.tostring and doc.formatDump. Not sure for now. Not sure 
that it matters. 
3. Use the objectified xml to populate internal state 
4. Look through the spec for validation rules beyond what the 
XSD is providing (e.g., sanity checks) 
5. Look into integration with biometrics
6. Fix fixtures/call_parameters.json 
7. Add support for session key generation and extraction of 
certificate information - done
8. Make this into a module 

Running 
-------

$ sudo apt-get install python-dev libxml2-dev libxslt1-dev libxmlsec1 libxmlsec1-dev 
$ sudo pip install lxml 
$ sudo pip install pyxmlsec 
$ python authrequest.py 
<Auth txn="" ac="public" xmlns="http://www.uidai.gov.in/authentication/uid-auth-request/1.0" ver="1.5" uid="123412341234" tid="" sa="public">
  <Skey ci="23233">ZWhoc2tz</Skey>
  <Uses pfa="n" bio="n" pin="n" pa="n" otp="n" pi="y"/>
  <Data>ZGZkc2ZkZmRz</Data>
</Auth>

The XML generated is XSD compliant
Validating this incoming XML
<?xml version="1.0"?> 
<Auth xmlns="http://www.uidai.gov.in/authentication/uid-auth-request/1.0" 
      ver="1.5" tid="public" ac="public" sa="public" 
      lk="MKg8njN6O+QRUmYF+TrbBUCqlrCnbN/Ns6hYbnnaOk99e5UGNhhE/xQ=" uid="999999990019" 
      txn="GEO.11051880"> 
      <Skey ci="20131003">Nc6DrZKFk1oQXxfgnFUl0mmtYYIPl0RGaFd2oINkpChU1++xdddMx6Dlbz6mEYs3 
            IyzChGjRXN5/al9r0runFX8LspTfMchwpxaaDIOyIUguBoYmPUqJDqTQcwey6Ntc 
            TJWFSgOvBg+omUkdbK/9GOQ5KWWrN+E0A9JN0IPU4IJqJZmsA6ETZlVoZteYtoMI 
            Ucv53qmxNPOEmJ3s4BC3ppHRRWRFMUp/eW7DFJ33W+uInZB6yekKE0dz8fYeo03w 
            2JUT1wlafL7aseb04nv5tNEbllHWafmbMpbv2pXKr+WPgytjrygt1LagGqF4a5Mr 
            /UTNwsy4m/YwlkWN0QcYVw== 
      </Skey> 
      <Uses otp="n" pin="n" bio="n" pa="n" pfa="n" pi="y" /> 
      <Data>YOn05vg5qMwElULpEmdiH0j6rM1XWcbQN0n+CFNQeazouCgjyPBH/a2SwbFgq/fF 
            CYUm+the8gQyYC36VO49NLcNcD7WdMhweoiDYgJoCX/t87Kbq/ABoAetfX7OLAck 
            /mHrTmw8tsfJgo4xGSzKZKr+pVn1O8dDHJjwgptySr7vp2Ntj6ogu6B905rsyTWw 
            73iMgoILDHf5soM3Pvde+/XW5rJD9AIPQGhHnKirwkiAgNIhtWU6ttYg4t6gHHbZ 
            0gVBwgjRzM3sDWKzK0EnmA== 
      </Data> 
      <Hmac>xy+JPoVN9dsWVm4YPZFwhVBKcUzzCTVvAxikT6BT5EcPgzX2JkLFDls+kLoNMpWe 
      </Hmac> 
</Auth> 

The XML generated is XSD compliant

$ python authresponse.py
<?xml version="1.0"?>
<AuthRes info="" txn="" code="-1" err="100" ts="2011-10-30T13:30:35" ret="n" xmlns="http://www.uidai.gov.in/authentication/uid-auth-response/1.0"/>

The XML generated is XSD compliant
Validating this incoming XML
<?xml version="1.0"?> 
<AuthRes  xmlns="http://www.uidai.gov.in/authentication/uid-auth-response/1.0"
	  ret="y" code="52" 
	  txn="322hfdjhsjkdhfjkds" err="100" info="" 
	  ts="2011-10-30T13:26:19"></AuthRes>

The XML generated is XSD compliant

$ python authrequest_signature.py fixtures/authrequest-with-sig.xml 
<?xml version="1.0"?>
<Auth xmlns="http://www.uidai.gov.in/authentication/uid-auth-request/1.0" ver="1.5" tid="public" ac="public" sa="public" lk="MKg8njN6O+QRUmYF+TrbBUCqlrCnbN/Ns6hYbnnaOk99e5UGNhhE/xQ=" uid="999999990019" txn="GEO.11051880"> 
....
R/l8w0hCInLusQMZeXgHcnxBGDSk1AQxKk5UfQmCwHNcRJMB5Zkj8+9n6T+/wx6D
tKDelktgIoo7w0EJ6MdVJ9Qzr5PJcYzX+ERgJEd/NNNVoPjFc2Al2odjToZdFN8+
/upJnBH02TRb1Wq63OtcuyBIFA==</X509Certificate>
</X509Data> 
            </KeyInfo> 
      </Signature> 
</Auth>

$ python authrequest_signature.py fixtures/authrequest-with-sig.xml  > new-authrequest-with-sig.xml
$ xmlsec --verify --pkcs12 fixtures/public.p12 --pwd public new-authrequest-with-sig.xml
func=xmlSecOpenSSLX509StoreVerify:file=x509vfy.c:line=360:obj=x509-store:subj=X509_verify_cert:error=4:crypto library function failed:subj=/C=IN/ST=KA/L=Bangalore/O=Public AUA/OU=Public/CN=Public AUA;err=20;msg=unable to get local issuer certificate
func=xmlSecOpenSSLX509StoreVerify:file=x509vfy.c:line=360:obj=x509-store:subj=X509_verify_cert:error=4:crypto library function failed:subj=/C=IN/ST=KA/L=Bangalore/O=Public AUA/OU=Public/CN=Public AUA;err=20;msg=unable to get local issuer certificate
func=xmlSecOpenSSLX509StoreVerify:file=x509vfy.c:line=408:obj=x509-store:subj=unknown:error=71:certificate verification failed:err=20;msg=unable to get local issuer certificate
OK
SignedInfo References (ok/all): 1/1
Manifests References (ok/all): 0/0

[Notes: The above error refers to the lack of CA certificate.] 

$ python auth_crypt.py 
certificate expiry =  Jun 28 04:40:44 2012 GMT
Encryption payload:  39jsjsfdhdshfd
Encrypted base64 encoded data:
hm8IKZIubO49F3y2RiLhBlW1tG3lAWep4j9l8rQ/XO/0OHKj4s+iehkkUw6Ew5KGes/yWeo993SRYw4/4sGT+fSNqCGw0LCL7WGdDKxuuoTuW0qytdfQQCydPICo1/fyy6RNl9n/v4+4eaf3UWgfg3oFq3d4J4cSDqfHC4ToCeQ=
Decrypted data
39jsjsfdhdshfd
Encrytion payload and decrypted data matched



CHANGES.txt
v0.1, 11/11/2011 -- Initial Release


authrequest.py
#
#<?xml version="1.0"?> 
#<Auth xmlns="http://www.uidai.gov.in/authentication/uid-auth-request/1.0" 
#      ver="1.5" tid="public" ac="public" sa="public" 
#      lk="MKg8njN6O+QRUmYF+TrbBUCqlrCnbN/Ns6hYbnnaOk99e5UGNhhE/xQ=" uid="999999990019" 
#      txn="GEO.11051880"> 
#      <Skey ci="20131003">Nc6DrZKFk...
#      </Skey> 
#      <Uses otp="n" pin="n" bio="n" pa="n" pfa="n" pi="y" /> 
#      <Data>YOn05vg5qMwElULpEmdiH0j6rM...
#      </Data> 
#      <Hmac>xy+JPoVN9dsWVm4YPZFwhVBKcUzzCTVvAxikT6BT5EcPgzX2JkLFDls+kLoNMpWe 
#      </Hmac> 
#</Auth> 
#

import sys
sys.path.append("lib") 

import libxml2
from lxml import etree, objectify 

import dumper 
import hashlib 
from config import Config 
import traceback 
import base64 
import random 
from datetime import datetime
from auth_crypt import AuthCrypt 
from M2Crypto import Rand 

class AuthRequest():

    """
    Base class to parse, validate, and generate auth requests going to
    the server. Mostly it will be used in the generate mode. The aim
    is to simplify writing applications around the auth request. We
    could potentially use this with AppEngine and Django that are
    python-based. This interface supports only v1.5 and public AuAs 
    """
    
    def __init__(self, cfg=None, biometrics=False, uid="", 
                 tid="", lk="", txn=""):
        
        self._cfg = cfg 
        self._biometrics = biometrics
        self._pidxml_biometrics = None
        self._pidxml_demographics = None 
        self._session_key = None

        self._tid = tid
        self._lk = lk
        self._ac = "public"
        self._ver = "1.5" 
        self._sa = "public" 
        self._uid = uid
        self._txn = txn 
        self._skey = { 
            '_ci': None, 
            '_text': None}
        self._uses  = { 
            '_otp': "n", 
            '_pin': "n",
            '_bio': "n", 
            '_pfa': "n",
            '_pi': "n",
            '_pa': "n",
            }
        self._hmac = ""
        self._data = ""

    def validate(self): 
        
        if ((self._skey['_ci'] == None) or (self._skey['_text'] == None)):
            raise Exception("Invalid Skey ci or text")
        
        if (self._pidxml_demographics == None and 
            self._pidxml_biometrics == None):
            raise Exception("Payload (demographics/biometics) not set") 

    def xsd_check(self, xml_text=None):
        
        if xml_text == None: 
            xml_text = self.tostring() 
        
        f = file(cfg.xsd.request)
        schema = etree.XMLSchema(file=f)
        parser = objectify.makeparser(schema = schema)
        try: 
            obj = objectify.fromstring(xml_text, parser)
            print "The XML generated is XSD compliant" 
        except: 
            print "[Error] Unable to parse incoming message" 
            traceback.print_exc(file=sys.stdout) 
            return None 
        return obj

    def set_skey(self):
        
        a = AuthCrypt(cfg.request.uid_cert_path, None) 
        when = a.get_cert_expiry() #Jun 28 04:40:44 2012 GMT
        expiry = datetime.strptime(when, "%b %d %H:%M:%S %Y %Z")
        self._session_key = Rand.rand_bytes(self._cfg.common.rsa_key_len) 
        print "session_key = ", self._session_key 
        self._skey['_ci'] = expiry.strftime("%Y%M%d")
        self._skey['_text'] = a.encrypt(self._session_key)

    def get_skey(self):
        return { 
            'ci': self._skey['_ci'],
            'text': self._skey['_text'],
            }

    def set_data(self, data=""):
        self._data = data 
        
    def get_data(self):
        self._data 

    def generate_xmldsig_template(self):
        "" 
        
    def set_hmac(self): 
        key = self._cfg.request.hmac_key 
        ""
        
    def set_pidxml_biometrics(self, datatype="FMR", data=None, ts=None):
        
        if (datatype != "FMR"): 
            raise Exception("Non FMR biometrics not supported") 
        
        if (data == None): 
            raise Exception("Data for biometrics inclusion is missing") 

        self._uses['_bio'] = "y"
        self._uses['_bt'] = "FMR"

        if ts == None:
            ts = Datetime.utcnow() 
        root = etree.Element('Pid', 
                             xmlns="http://www.uidai.gov.in/authentication/uid-auth-request-data/1.0",
                             ts=ts.strftime("%Y-%m-%dT%H:%M:%S"),
                             ver="1.0")
        bios = etree.SubElement(root, "Bios")
        bio=etree.SubElement(bios, "Bio", type="FMR")
        bio.text = data 
        doc = etree.ElementTree(root) 
        self._pidxml_biometrics = etree.tostring(doc,pretty_print=True)

    def set_pidxml_demographics(self, datatype="Name", data=None, ts=None):
        
        if (datatype != "Name" or data == None):
            raise Exception("Does not support demographic checks other than Name") 
        
        self._uses['_pi'] = "y" 
        
        if ts == None:
            ts = datetime.utcnow() 

        root = etree.Element('Pid', 
                             xmlns="http://www.uidai.gov.in/authentication/uid-auth-request-data/1.0",
                             ts=ts.strftime("%Y-%m-%dT%H:%M:%S"),
                             ver="1.0")
        demo = etree.SubElement(root, "Demo")
        pi=etree.SubElement(demo, "Pi", ms="E", name=data)
        doc = etree.ElementTree(root) 
        self._pidxml_demographics = etree.tostring(doc,pretty_print=True)

    def tostring(self):

        self.validate()

        root = etree.Element('Auth', 
                                xmlns="http://www.uidai.gov.in/authentication/uid-auth-request/1.0",
                                ver=self._ver,
                                tid=self._tid, 
                                ac=self._ac, 
                                sa=self._sa,
                                txn = self._txn,
                                uid = self._uid,
                                )
        skey = etree.SubElement(root, "Skey", ci=self._skey['_ci'])
        skey.text = base64.b64encode(self._skey['_text'])
        
        uses = etree.SubElement(root, "Uses", 
                                otp=self._uses['_otp'],
                                pin=self._uses['_pin'],
                                bio=self._uses['_bio'],
                                pfa=self._uses['_pfa'],
                                pi=self._uses['_pi'],
                                pa=self._uses['_pa'])
        
        data = etree.SubElement(root, "Data")
        data.text = base64.b64encode(self._data)

        doc = etree.ElementTree(root) 
        return ("<?xml version=\"1.0\"?>\n%s" %(etree.tostring(doc, pretty_print=True)))

    def load(self, xmlfile):        
        xml_text = file(xmlfile).read() 
        o = self.xsd_check(xml_text) 
        
if __name__ == '__main__':
    
    cfg = Config('auth.cfg') 
    x = AuthRequest(cfg, uid="123412341234", lk=cfg.common.license_key)
    x.set_skey() 
    x.set_pidxml_demographics(data="KKKK")
    x.set_data("dfdsfdfds") 
    s = x.tostring() 
    print s 
    x.xsd_check()

    test_xml = file('fixtures/authrequest.xml').read() 
    print "Validating this incoming XML" 
    print test_xml
    x.xsd_check(test_xml) 
    
